macro/R/continuous_step.R

```diff
@@ -1,51 +1,333 @@
+#' doesn't call this function.
+#'
+#' @param transition A list of transitions
+#' @param individual A list with individual properties
+#' @param when A time
+#' @param variables Any global simulation state
+#' @return logical whether transition can fire given this state
+#' @seealso \code{\link{when}}, \code{\link{fire}}
+#' @export
+is_enabled <- function(transition, individual, when, variables = NULL) {
+  transition_names <- names(transition)
+  enable_function <- ifelse(is.null(transition_names), transition[[1]], transition[["is_enabled"]])
+  if (!is.null(variables)) {
+    environment(enable_function) <- variables
+  }  # else don't need to create an environment
+  enable_function(individual, when)
+}
+
+
+#' Find when this transition will fire.
+#'
+#' This function is for testing the transitions on data for a
+#' single individual. It's a helper function. The running simulation
+#' doesn't call this function.
+#'
+#' @param transition A list of transitions
+#' @param individual A list with individual properties
+#' @param when A time
+#' @param variables Any global simulation state
+#' @return numeric time relative to now, when it would fire
+#' @seealso \code{\link{is_enabled}}, \code{\link{fire}}
+#' @export
+when <- function(transition, individual, when, variables = NULL) {
+  transition_names <- names(transition)
+  when_function <- ifelse(is.null(transition_names), transition[[2]], transition[["when"]])
+  if (!is.null(variables)) {
+    environment(when_function) <- variables
+  }  # else don't need to create an environment
+  when_function(individual, when)
+}
+
+
+#' Fire transition, changing the individual
+#'
+#' This function is for testing the transitions on data for a
+#' single individual. It's a helper function. The running simulation
+#' doesn't call this function.
+#'
+#' @param transition A list of transitions
+#' @param individual A list with individual properties
+#' @param when A time
+#' @param variables Any global simulation state
+#' @return a new list describing the individual
+#' @seealso \code{\link{is_enabled}}, \code{\link{when}}
+#' @export
+fire <- function(transition, individual, when, variables = NULL) {
+  transition_names <- names(transition)
+  fire_function <- ifelse(is.null(transition_names), transition[[3]], transition[["fire"]])
+  if (!is.null(variables)) {
+    environment(fire_function) <- variables
+  }  # else don't need to create an environment
+  fire_function(individual, when)
+}
+
+
+#' Set up firing times given an initial state.
+#'
+#' @param individuals a data frame of individuals, including columns for times.
+#' @param transitions a list of transitions for enabling and firing times
+#' @return a data frame of individuals, with times filled in
+#' @export
+initialize_times <- function(individuals, transitions) {
+  is_enabled <- transitions$is_enabled
+  when <- transitions$when
+  curtime <- 0
   for (person_idx in 1:nrow(individuals)) {
     state <- as.list(individuals[person_idx])
```

Comment:
this may be a bug; it pulls out the first column of the `data.table` object called "individuals" for _all_ individuals. maybe `as.list(individuals[person_idx,])`, assuming that the state is column one?

Reply:
Seems to work?
```
> dt <- data.table(who = 1:3, when = c(0.3, 0.1, 0.7))
> as.list(dt[1])
$who
[1] 1

$when
[1] 0.3
```

Reply:
You could argue that I shouldn't convert it to a list, that this will slow the calculation.

Reply:
huh! ok. I'm obviously not as familiar with data.table as I thought I was.

Reply:
I'm thoroughly confused at this point. My examples do one thing, and the code does another. Humph. I fixed it, and it's now running.